File: scripts/target_models/helpers/cusum.py
"""
CUSUM (Cumulative Sum Control Chart) Helper.

Change point detection using CUSUM with:
- Rolling normalization (adaptive to changing conditions)
- Reset after detection (prevents runaway accumulation)
- Minimum spacing (prevents clustered detections)
- Multi-target support (returns, volatility)

Features Generated (~12 per target-horizon):
- cusum_pos/neg: Raw CUSUM values
- changepoint_up/down/any: Binary change indicators
- magnitude: Size of detected change
- days_since: Time since last changepoint
- combined: Either vol or return changepoint

Now uses Rust backend for ~200x speedup when available.
"""

from dataclasses import dataclass
from typing import Any

import numpy as np

from .base import BaseHelper, HelperConfig

# Try to import Rust backend
try:
    import riskyield_rust as _rust

    HAS_RUST = True
except ImportError:
    HAS_RUST = False


# =============================================================================
# CONFIGURATION
# =============================================================================
@dataclass
class CUSUMConfig(HelperConfig):
    """Configuration specific to CUSUM helper."""

    # CUSUM parameters
    threshold: float = 2.0  # Detection threshold (in std devs)
    drift: float = 0.0  # Drift term (0 = no drift adjustment)
    min_spacing: int = 5  # Minimum bars between detections

    # Rolling normalization
    rolling_window: int = 63  # Window for z-score normalization

    # Feature columns to use (by name pattern or index)
    # If empty, will use returns and volatility proxies
    return_col_idx: int = 0  # Index of return-like feature
    vol_col_idx: int = 1  # Index of volatility-like feature


# =============================================================================
# CUSUM HELPER
# =============================================================================
class CUSUMHelper(BaseHelper):
    """
    CUSUM change point detection helper.

    Applies CUSUM algorithm to returns and volatility series to
    detect regime changes and structural breaks.

    Uses Rust backend when available for ~200x speedup.

    Features generated:
    - H_{prefix}_cusum_ret_pos: Positive CUSUM for returns
    - H_{prefix}_cusum_ret_neg: Negative CUSUM for returns
    - H_{prefix}_cusum_vol_pos: Positive CUSUM for volatility
    - H_{prefix}_cusum_vol_neg: Negative CUSUM for volatility
    - H_{prefix}_cp_ret_up: Return increase changepoint
    - H_{prefix}_cp_ret_down: Return decrease changepoint
    - H_{prefix}_cp_vol_up: Volatility increase changepoint
    - H_{prefix}_cp_vol_down: Volatility decrease changepoint
    - H_{prefix}_cp_any: Any changepoint detected
    - H_{prefix}_cp_magnitude: Change magnitude (if detected)
    - H_{prefix}_days_since_cp: Days since last changepoint
    - H_{prefix}_cp_count_21: Rolling 21-bar changepoint count
    """

    def __init__(self, config: CUSUMConfig | None = None):
        super().__init__(config or CUSUMConfig())
        self.config: CUSUMConfig  # Type hint

        # Fitted parameters (rolling stats from training)
        self._return_mean: float = 0.0
        self._return_std: float = 1.0
        self._vol_mean: float = 0.0
        self._vol_std: float = 1.0

    @property
    def supports_incremental(self) -> bool:
        """CUSUM requires full refit for threshold calibration.

        The detection thresholds need to be calibrated on the full
        training data distribution to maintain proper false positive rates.
        """
        return False

    @property
    def helper_name(self) -> str:
        return "cusum"

    def _fit_impl(self, X: np.ndarray, y: np.ndarray | None = None) -> None:
        """
        Compute baseline statistics for CUSUM normalization.

        Args:
            X: Feature matrix (n_samples, n_features)
            y: Ignored
        """
        n_samples, n_features = X.shape

        # Extract return and volatility proxies
        ret_idx = min(self.config.return_col_idx, n_features - 1)
        vol_idx = min(self.config.vol_col_idx, n_features - 1)

        returns = X[:, ret_idx]
        volatility = X[:, vol_idx]

        # Compute baseline statistics (robust to outliers)
        self._return_mean = float(np.nanmedian(returns))
        self._return_std = float(np.nanstd(returns)) + 1e-8

        self._vol_mean = float(np.nanmedian(volatility))
        self._vol_std = float(np.nanstd(volatility)) + 1e-8

        self._fit_params = {
            "n_samples": n_samples,
            "return_col_idx": ret_idx,
            "vol_col_idx": vol_idx,
            "return_mean": self._return_mean,
            "return_std": self._return_std,
            "vol_mean": self._vol_mean,
            "vol_std": self._vol_std,
        }

        if self.config.verbose:
            print(f"CUSUMHelper fitted on {X.shape}")
            print(
                f"  Return baseline: μ={self._return_mean:.6f}, σ={self._return_std:.6f}"
            )
            print(f"  Vol baseline: μ={self._vol_mean:.6f}, σ={self._vol_std:.6f}")

    def _transform_impl(self, X: np.ndarray) -> np.ndarray:
        """
        Generate CUSUM-based features.

        Args:
            X: Feature matrix (n_samples, n_features)

        Returns:
            Feature matrix (n_samples, 12) with CUSUM features
        """
        n_features = X.shape[1]

        # Extract series
        ret_idx = min(self.config.return_col_idx, n_features - 1)
        vol_idx = min(self.config.vol_col_idx, n_features - 1)

        returns = X[:, ret_idx]
        volatility = X[:, vol_idx]

        # Use Rust backend if available (~200x faster)
        if HAS_RUST:
            return self._transform_rust(returns, volatility)

        # Fallback to Python implementation
        return self._transform_python(returns, volatility)

    def _transform_rust(
        self, returns: np.ndarray, volatility: np.ndarray
    ) -> np.ndarray:
        """Rust-accelerated CUSUM transform."""
        # Ensure contiguous arrays
        returns = np.ascontiguousarray(returns, dtype=np.float64)
        volatility = np.ascontiguousarray(volatility, dtype=np.float64)

        # Call Rust backend
        features = _rust.py_cusum_transform(
            returns,
            volatility,
            self._return_mean,
            self._return_std,
            self._vol_mean,
            self._vol_std,
            threshold=self.config.threshold,
            drift=self.config.drift,
            min_spacing=self.config.min_spacing,
            rolling_window=self.config.rolling_window,
        )

        return features

    def _transform_python(
        self, returns: np.ndarray, volatility: np.ndarray
    ) -> np.ndarray:
        """Pure Python CUSUM transform (fallback)."""
        # Compute CUSUM for returns
        ret_cusum_pos, ret_cusum_neg, ret_cp_up, ret_cp_down = self._compute_cusum(
            returns,
            self._return_mean,
            self._return_std,
        )

        # Compute CUSUM for volatility
        vol_cusum_pos, vol_cusum_neg, vol_cp_up, vol_cp_down = self._compute_cusum(
            volatility,
            self._vol_mean,
            self._vol_std,
        )

        # Combined features
        any_cp = np.maximum.reduce([ret_cp_up, ret_cp_down, vol_cp_up, vol_cp_down])

        # Change magnitude (absolute z-score at changepoint)
        ret_zscore = np.abs((returns - self._return_mean) / self._return_std)
        vol_zscore = np.abs((volatility - self._vol_mean) / self._vol_std)
        magnitude = np.where(any_cp > 0, np.maximum(ret_zscore, vol_zscore), 0.0)

        # Days since last changepoint
        days_since = self._compute_days_since(any_cp)

        # Rolling changepoint count (21-bar window)
        cp_count_21 = self._rolling_sum(any_cp, 21)

        # Stack all features
        features = np.column_stack(
            [
                ret_cusum_pos,
                ret_cusum_neg,
                vol_cusum_pos,
                vol_cusum_neg,
                ret_cp_up,
                ret_cp_down,
                vol_cp_up,
                vol_cp_down,
                any_cp,
                magnitude,
                days_since,
                cp_count_21,
            ]
        )

        return features

    def _compute_cusum(
        self,
        series: np.ndarray,
        mean: float,
        std: float,
    ) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """
        Compute CUSUM with reset and minimum spacing.

        Args:
            series: Input time series
            mean: Baseline mean
            std: Baseline std

        Returns:
            cusum_pos, cusum_neg, changepoint_up, changepoint_down
        """
        n = len(series)
        threshold = self.config.threshold
        drift = self.config.drift
        min_spacing = self.config.min_spacing
        window = self.config.rolling_window

        cusum_pos = np.zeros(n)
        cusum_neg = np.zeros(n)
        cp_up = np.zeros(n)
        cp_down = np.zeros(n)

        last_cp_idx = -min_spacing  # Allow first detection

        for i in range(1, n):
            # Use rolling normalization if enough history
            if window > 0 and i >= window:
                window_data = series[max(0, i - window) : i]
                local_mean = np.nanmean(window_data)
                local_std = np.nanstd(window_data) + 1e-8
            else:
                local_mean = mean
                local_std = std

            # Z-score
            z = (series[i] - local_mean) / local_std

            # CUSUM update
            cusum_pos[i] = max(0, cusum_pos[i - 1] + z - drift)
            cusum_neg[i] = min(0, cusum_neg[i - 1] + z + drift)

            # Detection with minimum spacing
            if i - last_cp_idx >= min_spacing:
                if cusum_pos[i] > threshold:
                    cp_up[i] = 1
                    cusum_pos[i] = 0  # Reset after detection
                    last_cp_idx = i
                elif cusum_neg[i] < -threshold:
                    cp_down[i] = 1
                    cusum_neg[i] = 0  # Reset after detection
                    last_cp_idx = i

        return cusum_pos, cusum_neg, cp_up, cp_down

    def _compute_days_since(self, changepoints: np.ndarray) -> np.ndarray:
        """Compute days since last changepoint."""
        n = len(changepoints)
        days_since = np.zeros(n)

        last_cp = -1
        for i in range(n):
            if changepoints[i] > 0:
                last_cp = i
            days_since[i] = i - last_cp if last_cp >= 0 else i

        return days_since

    def _rolling_sum(self, arr: np.ndarray, window: int) -> np.ndarray:
        """Compute rolling sum."""
        n = len(arr)
        result = np.zeros(n)

        cumsum = np.cumsum(arr)
        result[window - 1 :] = cumsum[window - 1 :] - np.concatenate(
            [[0], cumsum[:-window]]
        )

        # Handle early values
        for i in range(min(window - 1, n)):
            result[i] = np.sum(arr[: i + 1])

        return result

    def _get_feature_names(self) -> list[str]:
        """Get feature names."""
        return [
            self._make_feature_name("cusum_ret_pos"),
            self._make_feature_name("cusum_ret_neg"),
            self._make_feature_name("cusum_vol_pos"),
            self._make_feature_name("cusum_vol_neg"),
            self._make_feature_name("cp_ret_up"),
            self._make_feature_name("cp_ret_down"),
            self._make_feature_name("cp_vol_up"),
            self._make_feature_name("cp_vol_down"),
            self._make_feature_name("cp_any"),
            self._make_feature_name("cp_magnitude"),
            self._make_feature_name("days_since_cp"),
            self._make_feature_name("cp_count_21"),
        ]

    def validate(
        self,
        X_val: np.ndarray,
        y_val: np.ndarray | None = None,
    ) -> dict[str, float]:
        """Compute validation metrics."""
        features = self._transform_impl(X_val)

        # Changepoint detection rate
        cp_any_idx = 8  # Index of cp_any feature
        cp_rate = np.mean(features[:, cp_any_idx])

        # Average days between changepoints
        days_since_idx = 10
        avg_spacing = np.mean(features[:, days_since_idx])

        return {
            "changepoint_rate": float(cp_rate),
            "avg_days_since_cp": float(avg_spacing),
        }

    def get_params(self) -> dict[str, Any]:
        """Get detailed parameters."""
        return {
            **super().get_params(),
            "threshold": self.config.threshold,
            "drift": self.config.drift,
            "min_spacing": self.config.min_spacing,
            "rolling_window": self.config.rolling_window,
        }


# =============================================================================
# FACTORY
# =============================================================================
def create_cusum_helper(
    target: str,
    horizon: int,
    task_type: str = "regression",
    **kwargs,
) -> CUSUMHelper:
    """
    Create CUSUMHelper for a specific target-horizon.

    Args:
        target: Target name
        horizon: Prediction horizon
        task_type: Task type
        **kwargs: Additional config overrides

    Returns:
        Configured CUSUMHelper
    """
    config = CUSUMConfig(
        target=target,
        horizon=horizon,
        task_type=task_type,
        prefix=f"{target[:3]}_{horizon}",
        **kwargs,
    )
    return CUSUMHelper(config)

File: riskyield_rust/src/cusum.rs
//! CUSUM (Cumulative Sum Control Chart) - Rust Implementation
//!
//! High-performance change point detection using CUSUM algorithm.
//! Provides ~200x speedup over Python implementation.
//!
//! Features generated (12 total):
//! - cusum_ret_pos/neg: Cumulative sums for returns
//! - cusum_vol_pos/neg: Cumulative sums for volatility
//! - cp_ret_up/down: Changepoint flags for returns
//! - cp_vol_up/down: Changepoint flags for volatility
//! - cp_any: Any changepoint detected
//! - cp_magnitude: Size of change at detection
//! - days_since_cp: Bars since last changepoint
//! - cp_count_21: Rolling 21-bar changepoint count

use rayon::prelude::*;

/// Configuration for CUSUM algorithm
#[derive(Clone, Debug)]
pub struct CUSUMConfig {
    pub threshold: f64,
    pub drift: f64,
    pub min_spacing: usize,
    pub rolling_window: usize,
}

impl Default for CUSUMConfig {
    fn default() -> Self {
        Self {
            threshold: 2.0,
            drift: 0.0,
            min_spacing: 5,
            rolling_window: 63,
        }
    }
}

/// Result of CUSUM computation for a single series
#[derive(Clone, Debug)]
pub struct CUSUMResult {
    pub cusum_pos: Vec<f64>,
    pub cusum_neg: Vec<f64>,
    pub cp_up: Vec<f64>,
    pub cp_down: Vec<f64>,
}

/// Full CUSUM features output (12 features)
#[derive(Clone, Debug)]
pub struct CUSUMFeatures {
    pub cusum_ret_pos: Vec<f64>,
    pub cusum_ret_neg: Vec<f64>,
    pub cusum_vol_pos: Vec<f64>,
    pub cusum_vol_neg: Vec<f64>,
    pub cp_ret_up: Vec<f64>,
    pub cp_ret_down: Vec<f64>,
    pub cp_vol_up: Vec<f64>,
    pub cp_vol_down: Vec<f64>,
    pub cp_any: Vec<f64>,
    pub cp_magnitude: Vec<f64>,
    pub days_since_cp: Vec<f64>,
    pub cp_count_21: Vec<f64>,
}

/// Compute rolling mean for a window ending at index i (exclusive)
#[inline]
fn rolling_mean(data: &[f64], end: usize, window: usize) -> f64 {
    if end < window {
        // Not enough data, use what we have
        let start = 0;
        let slice = &data[start..end];
        if slice.is_empty() {
            return 0.0;
        }
        let mut sum = 0.0;
        let mut count = 0;
        for &v in slice {
            if v.is_finite() {
                sum += v;
                count += 1;
            }
        }
        if count == 0 { 0.0 } else { sum / count as f64 }
    } else {
        let start = end - window;
        let slice = &data[start..end];
        let mut sum = 0.0;
        let mut count = 0;
        for &v in slice {
            if v.is_finite() {
                sum += v;
                count += 1;
            }
        }
        if count == 0 { 0.0 } else { sum / count as f64 }
    }
}

/// Compute rolling std for a window ending at index i (exclusive)
#[inline]
fn rolling_std(data: &[f64], end: usize, window: usize) -> f64 {
    let mean = rolling_mean(data, end, window);
    
    let (start, n) = if end < window {
        (0, end)
    } else {
        (end - window, window)
    };
    
    if n < 2 {
        return 1e-8;
    }
    
    let slice = &data[start..end];
    let mut sum_sq = 0.0;
    let mut count = 0;
    for &v in slice {
        if v.is_finite() {
            let diff = v - mean;
            sum_sq += diff * diff;
            count += 1;
        }
    }
    
    if count < 2 {
        1e-8
    } else {
        (sum_sq / count as f64).sqrt().max(1e-8)
    }
}

/// Core CUSUM algorithm for a single series
/// 
/// Computes cumulative sums with threshold detection and reset.
pub fn compute_cusum_single(
    series: &[f64],
    baseline_mean: f64,
    baseline_std: f64,
    config: &CUSUMConfig,
) -> CUSUMResult {
    let n = series.len();
    let mut cusum_pos = vec![0.0; n];
    let mut cusum_neg = vec![0.0; n];
    let mut cp_up = vec![0.0; n];
    let mut cp_down = vec![0.0; n];
    
    let threshold = config.threshold;
    let drift = config.drift;
    let min_spacing = config.min_spacing;
    let window = config.rolling_window;
    
    let mut last_cp_idx: i64 = -(min_spacing as i64);
    
    for i in 1..n {
        // Use rolling normalization if enough history, else baseline
        let (local_mean, local_std) = if window > 0 && i >= window {
            (rolling_mean(series, i, window), rolling_std(series, i, window))
        } else {
            (baseline_mean, baseline_std)
        };
        
        // Z-score
        let z = if local_std > 1e-10 {
            (series[i] - local_mean) / local_std
        } else {
            0.0
        };
        
        // CUSUM update
        cusum_pos[i] = (cusum_pos[i - 1] + z - drift).max(0.0);
        cusum_neg[i] = (cusum_neg[i - 1] + z + drift).min(0.0);
        
        // Detection with minimum spacing
        let since_last = i as i64 - last_cp_idx;
        if since_last >= min_spacing as i64 {
            if cusum_pos[i] > threshold {
                cp_up[i] = 1.0;
                cusum_pos[i] = 0.0; // Reset after detection
                last_cp_idx = i as i64;
            } else if cusum_neg[i] < -threshold {
                cp_down[i] = 1.0;
                cusum_neg[i] = 0.0; // Reset after detection
                last_cp_idx = i as i64;
            }
        }
    }
    
    CUSUMResult {
        cusum_pos,
        cusum_neg,
        cp_up,
        cp_down,
    }
}

/// Compute days since last changepoint
fn compute_days_since(changepoints: &[f64]) -> Vec<f64> {
    let n = changepoints.len();
    let mut days_since = vec![0.0; n];
    
    let mut last_cp: i64 = -1;
    for i in 0..n {
        if changepoints[i] > 0.0 {
            last_cp = i as i64;
        }
        days_since[i] = if last_cp >= 0 {
            (i as i64 - last_cp) as f64
        } else {
            i as f64
        };
    }
    
    days_since
}

/// Compute rolling sum (efficient O(n) implementation)
fn rolling_sum(arr: &[f64], window: usize) -> Vec<f64> {
    let n = arr.len();
    let mut result = vec![0.0; n];
    
    // Compute cumsum
    let mut cumsum = vec![0.0; n + 1];
    for i in 0..n {
        cumsum[i + 1] = cumsum[i] + arr[i];
    }
    
    // Rolling sum using cumsum difference
    for i in 0..n {
        if i + 1 >= window {
            result[i] = cumsum[i + 1] - cumsum[i + 1 - window];
        } else {
            result[i] = cumsum[i + 1];
        }
    }
    
    result
}

/// Main CUSUM transform function
/// 
/// Computes all 12 CUSUM features for returns and volatility series.
///
/// # Arguments
/// * `returns` - Returns series (column 0)
/// * `volatility` - Volatility series (column 1)
/// * `ret_mean` - Baseline mean for returns (from fit)
/// * `ret_std` - Baseline std for returns (from fit)
/// * `vol_mean` - Baseline mean for volatility (from fit)
/// * `vol_std` - Baseline std for volatility (from fit)
/// * `config` - CUSUM configuration
///
/// # Returns
/// CUSUMFeatures with all 12 feature vectors
pub fn cusum_transform(
    returns: &[f64],
    volatility: &[f64],
    ret_mean: f64,
    ret_std: f64,
    vol_mean: f64,
    vol_std: f64,
    config: &CUSUMConfig,
) -> CUSUMFeatures {
    let n = returns.len();
    
    // Compute CUSUM for returns
    let ret_result = compute_cusum_single(returns, ret_mean, ret_std, config);
    
    // Compute CUSUM for volatility
    let vol_result = compute_cusum_single(volatility, vol_mean, vol_std, config);
    
    // Combined: any changepoint
    let mut cp_any = vec![0.0; n];
    for i in 0..n {
        if ret_result.cp_up[i] > 0.0 
            || ret_result.cp_down[i] > 0.0 
            || vol_result.cp_up[i] > 0.0 
            || vol_result.cp_down[i] > 0.0 
        {
            cp_any[i] = 1.0;
        }
    }
    
    // Compute z-scores for magnitude calculation
    let mut cp_magnitude = vec![0.0; n];
    for i in 0..n {
        if cp_any[i] > 0.0 {
            let ret_zscore = ((returns[i] - ret_mean) / ret_std.max(1e-10)).abs();
            let vol_zscore = ((volatility[i] - vol_mean) / vol_std.max(1e-10)).abs();
            cp_magnitude[i] = ret_zscore.max(vol_zscore);
        }
    }
    
    // Days since last changepoint
    let days_since_cp = compute_days_since(&cp_any);
    
    // Rolling 21-bar changepoint count
    let cp_count_21 = rolling_sum(&cp_any, 21);
    
    CUSUMFeatures {
        cusum_ret_pos: ret_result.cusum_pos,
        cusum_ret_neg: ret_result.cusum_neg,
        cusum_vol_pos: vol_result.cusum_pos,
        cusum_vol_neg: vol_result.cusum_neg,
        cp_ret_up: ret_result.cp_up,
        cp_ret_down: ret_result.cp_down,
        cp_vol_up: vol_result.cp_up,
        cp_vol_down: vol_result.cp_down,
        cp_any,
        cp_magnitude,
        days_since_cp,
        cp_count_21,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cusum_basic() {
        let series = vec![0.0, 0.1, 0.2, 0.3, 0.5, 0.8, 1.0, 0.9, 0.7, 0.5];
        let config = CUSUMConfig {
            threshold: 2.0,
            drift: 0.0,
            min_spacing: 2,
            rolling_window: 0, // Use baseline only
        };
        
        let result = compute_cusum_single(&series, 0.5, 0.3, &config);
        
        assert_eq!(result.cusum_pos.len(), series.len());
        assert_eq!(result.cusum_neg.len(), series.len());
        assert_eq!(result.cusum_pos[0], 0.0); // First value always 0
    }

    #[test]
    fn test_rolling_sum() {
        let arr = vec![1.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0];
        let result = rolling_sum(&arr, 3);
        
        assert_eq!(result[0], 1.0); // Just first element
        assert_eq!(result[1], 1.0); // First two elements
        assert_eq!(result[2], 2.0); // [1, 0, 1]
        assert_eq!(result[3], 1.0); // [0, 1, 0]
    }

    #[test]
    fn test_days_since() {
        let cp = vec![0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0];
        let result = compute_days_since(&cp);
        
        assert_eq!(result[0], 0.0); // No CP yet
        assert_eq!(result[1], 1.0);
        assert_eq!(result[2], 0.0); // CP at index 2
        assert_eq!(result[3], 1.0);
        assert_eq!(result[4], 2.0);
        assert_eq!(result[5], 0.0); // CP at index 5
        assert_eq!(result[6], 1.0);
    }
}
